src/main/org/deidentifier/arx/ShadowModelMembershipRisk.java

```diff
@@ -336,30 +376,627 @@ private Data getCopy(DataHandle handle, int[] sample) {
+         * 
+         * @return
+         */
+        private double[] calculateHistogramFeatures() {
+        
+            // Initialize array used to store the parts of the result
+            double[][] result = new double[columns.length][];
+            
+            for (int i = 0; i < columns.length; i++) {
+
+                // Obtain attribute details
+                int c = columns[i];
+                String attributeName = handle.getAttributeName(c);
+                DataType<?> _type = handle.getDefinition().getDataType(attributeName);
+                Class<?> _clazz = _type.getDescription().getWrappedClass();
+                
+                if (_clazz.equals(Double.class)) {
+                    
+                    double min = metaData.getMinValue(i);
+                    double max = metaData.getMaxValue(i);
+                    double binSize = (max - min) / NUM_BINS;
+                    
+                    double[] freqs = new double[NUM_BINS];
+                    Double[] values = getColumnAsDouble(c);
+                    
+                    /*
+                    for(int d = 0; d < NUM_BINS; d++) {
+                        System.out.println(d + " | " + (d*binSize+min) + " | " );
+                    }
+                    */
+                    
+                    for(Double v : values) {
+                        // calculate bin
+                        int bin = (int) ((v - min) / binSize);
+                        // check range
+                        if(0 > bin || bin > NUM_BINS) {
+                            throw new RuntimeException("Value out of histogram range");
+                        }
+                        // TODO Dirty quick-fix
+                        if(bin == NUM_BINS) {
+                            bin -= 1;
+                        }
+                        // inc freq of bin
+                        freqs[bin] += 1d;
+                    }
+                    
+                    result[i] = freqs;
+                    
+                } else if (_clazz.equals(String.class)) {
+                    
+                    // Count frequencies of values
+                    int column = columns[i];
+                    IntIntOpenHashMap map = new IntIntOpenHashMap();
+                    for (int row = 0; row < handle.getNumRows(); row++) {
+                        int code = handle.internalGetEncodedValue(row, column, false); // Beware that code can be -1
+                        map.putOrAdd(code, 1, 1);
+                    }
+                    
+                    int[] availableValues = metaData.getAvialableValues(i);
+                    double[] freqs = new double[availableValues.length];
+                    
+                    // Count freqs of all available values.
+                    for(int v = 0; v < availableValues.length; v++) {
+                        int value = availableValues[v];
```

Comment:
This whole block can be replaced with freqs[v] = map.getOrDefault(availableValues[v], 0d);

Reply:
Thanks! Done.